### PR TITLE
feat: add built-in udfs in config

### PR DESCRIPTION
**Pull request by BohuTANG**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

This PR:
*  spill out IDM config type conversion from the config crate, make the config only define as string
* `UserConfig` to meta `UserAuthInfo` convert in `builtin/builtin_user.rs`
* `UDFConfig` to  meta `UserDefinedFunction` convert in `builtin/builtin_udf.rs`

## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15938)

### CODE REVIEW

File: src/query/config/src/config.rs [hunk 1]
```diff
@@ -1805,7 +1818,8 @@ impl From<InnerQueryConfig> for QueryConfig {
             jwt_key_files: inner.jwt_key_files,
             default_storage_format: inner.default_storage_format,
             default_compression: inner.default_compression,
-            users: users_from_inner(inner.idm.users),
+            users: inner.idm.users,
+            udfs: inner.idm.udfs,
```
Comment (everpcpc) [inline src/query/config/src/config.rs:1822]: Since `idm` means `identity manager`, maybe `udfs` should be locationed outside of it?

Comment (BohuTANG) [reply src/query/config/src/config.rs:1822]: I plan to change the name from `idm` to `builtin`.

**Comment by BohuTANG** [comment]
Will fix the failed test later.

File: src/query/config/src/config.rs (hunk 1 above)
Comment (youngsofun) [reply src/query/config/src/config.rs]: what about `predefined`  or `preconfigured`?  `built-in` sounds like `hard-coded` in code?

File: src/query/service/src/builtin/builtin_udfs.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    pub async fn to_meta_udfs(&self) -> Result<HashMap<String, UserDefinedFunction>> {
+        let mut udf_map = HashMap::new();
+
+        for udf_config in self.udf_configs.iter() {
+            match self
+                .parse_udf_definition(&udf_config.name, &udf_config.definition)
+                .await
+            {
+                Ok(user_defined_function) => {
+                    udf_map.insert(user_defined_function.name.clone(), user_defined_function);
```
Comment (youngsofun) [inline src/query/service/src/builtin/builtin_udfs.rs:95]: check for dup names here?

File: src/query/service/src/builtin/builtin_udfs.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    udf_configs: Vec<UDFConfig>,
+}
+
+impl BuiltinUDFs {
+    pub fn create(udf_configs: Vec<UDFConfig>) -> BuiltinUDFs {
+        BuiltinUDFs { udf_configs }
+    }
+
+    // Parse the UDF definition and return the UserDefinedFunction.
+    async fn parse_udf_definition(
```
Comment (youngsofun) [inline src/query/service/src/builtin/builtin_udfs.rs:40]: not need to be async?

File: src/query/service/src/global_services.rs [hunk 4]
```diff
@@ -107,13 +110,25 @@ impl GlobalServices {
...
-        .await?;
+
+        // Init user manager.
+        // Builtin users and udfs are created here.
+        {
+            let built_in_users = BuiltinUsers::create(config.query.builtin.users.clone());
+            let built_in_udfs = BuiltinUDFs::create(config.query.builtin.udfs.clone());
+            let builtin = BuiltIn {
+                users: built_in_users.to_meta_auth_infos()?,
+                udfs: built_in_udfs.to_meta_udfs().await?,
```
Comment (youngsofun) [inline src/query/service/src/global_services.rs:121]: Provide additional context for this error, so the user can avoid encountering a parse error without knowing its origin.

File: src/query/service/src/builtin/builtin_udfs.rs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                ))),
+            },
+            _ => Err(ErrorCode::SyntaxException(format!(
+                "Invalid built-in UDF definition: '{}', expected CreateUDF but got: {:?}",
+                definition, stmt
+            ))),
+        }
+    }
+
+    pub async fn to_meta_udfs(&self) -> Result<HashMap<String, UserDefinedFunction>> {
```
Comment (youngsofun) [inline src/query/service/src/builtin/builtin_udfs.rs:86]: why call it `to_meta_xxx?`

File: src/query/sql/src/planner/semantic/type_check.rs [hunk 6]
```diff
@@ -2176,14 +2176,14 @@ impl<'a> TypeChecker<'a> {
...
-            .set_span(query_scalar.span()));
+                .set_span(query_scalar.span()));
         };
         let Some(query_text) = query_expr.value.as_string() else {
             return Err(ErrorCode::SemanticError(format!(
                 "invalid arguments for search function, query text must be a constant string, but got {}",
                 query_arg
             ))
-            .set_span(query_scalar.span()));
+                .set_span(query_scalar.span()));
```
Comment (youngsofun) [inline src/query/sql/src/planner/semantic/type_check.rs:2186]: there are many formatting changes in this?
it seems not what u intended from your commit messages.
and ci-check not complaint about them.
is this expected?
will it cause these lines to be changed back and forth in prs?

File: src/query/service/src/builtin/builtin_udfs.rs (hunk 2 above)
Comment (BohuTANG) [reply src/query/service/src/builtin/builtin_udfs.rs:90]: If the built-in udfs config not invalid, the global init will log a error and skip, duplicate is no need check here.

File: src/query/service/src/builtin/builtin_udfs.rs (hunk 3 above)
Comment (BohuTANG) [reply src/query/service/src/builtin/builtin_udfs.rs]: Great, thanks for remind.

File: src/query/service/src/global_services.rs (hunk 4 above)
Comment (BohuTANG) [reply src/query/service/src/global_services.rs]: There is no need to raise errors when the global init, if the udf&user config is invalid, we should skip them and leave a error.

File: src/query/service/src/builtin/builtin_udfs.rs (hunk 5 above)
Comment (BohuTANG) [reply src/query/service/src/builtin/builtin_udfs.rs]: Hmm, make the name better now.

File: src/query/sql/src/planner/semantic/type_check.rs (hunk 6 above)
Comment (BohuTANG) [reply src/query/sql/src/planner/semantic/type_check.rs:2186]: Weird.
Here I respect: `cargo fmt --all` with  rustfmt.toml.

File: src/query/config/src/config.rs (hunk 1 above)
Comment (youngsofun) [reply src/query/config/src/config.rs]: @BohuTANG

**Comment by youngsofun** [comment]
what about predefined or preconfigured? built-in sounds like hard-coded in code?

**Comment by BohuTANG** [comment]
> what about `predefined` or `preconfigured`? `built-in` sounds like hard-coded in code?

This PR make the the built-in users/udfs configured in the config file not code, built-in is make sense I think.

**Comment by BohuTANG** [comment]
The unit-test failed not releated to this PR:
```
--- STDOUT:              databend-common-meta-api txn_backoff::tests::test_backoff ---

running 1 test
test txn_backoff::tests::test_backoff ... FAILED

failures:

failures:
    txn_backoff::tests::test_backoff

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 5 filtered out; finished in 0.19s


--- STDERR:              databend-common-meta-api txn_backoff::tests::test_backoff ---
thread 'txn_backoff::tests::test_backoff' panicked at src/meta/api/src/txn_backoff.rs:156:9:
0.088934727 is expected to be 2 + 5 + 10 + 14 + 20 milliseconds
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by BohuTANG** [comment]
Will merge for bendml with the torrmow release. Any suggestions also welcome.

**Pull request merged**